File: tests/py/test_cross_platform_basic_types_serialization.py
#!/usr/bin/env python3
"""
Test script for cross-platform basic types serialization - writes test data to file.
This test populates a BasicTypesMessage from expected_values.json and writes it to a binary file.
"""

import sys
import os
import json


def print_failure_details(label, expected_values=None, actual_values=None, raw_data=None):
    """Print detailed failure information"""
    print(f"\n{'='*60}")
    print(f"FAILURE DETAILS: {label}")
    print(f"{'='*60}")
    
    if expected_values:
        print("\nExpected Values:")
        for key, val in expected_values.items():
            print(f"  {key}: {val}")
    
    if actual_values:
        print("\nActual Values:")
        for key, val in actual_values.items():
            print(f"  {key}: {val}")
    
    if raw_data:
        print(f"\nRaw Data ({len(raw_data)} bytes):")
        print(f"  Hex: {raw_data.hex()}")
    
    print(f"{'='*60}\n")


def load_expected_values():
    """Load expected values from JSON file"""
    json_path = os.path.join(os.path.dirname(__file__), '..', 'expected_values.json')
    try:
        with open(json_path, 'r') as f:
            data = json.load(f)
        return data['basic_types']
    except Exception as e:
        print(f"Error loading expected values: {e}")
        return None


def create_test_data():
    """Create test data for cross-platform compatibility testing"""
    try:
        sys.path.insert(0, '../generated/py')
        from basic_types_sf import BasicTypesBasicTypesMessage
        from struct_frame_parser import BasicFrame

        # Load expected values from JSON
        expected = load_expected_values()
        if not expected:
            return False

        # Handle large_uint which may be a string in JSON due to JavaScript number limits
        large_uint = expected['large_uint']
        if isinstance(large_uint, str):
            large_uint = int(large_uint)

        # Create test message using values from JSON
        msg = BasicTypesBasicTypesMessage(
            small_int=expected['small_int'],
            medium_int=expected['medium_int'],
            regular_int=expected['regular_int'],
            large_int=expected['large_int'],
            small_uint=expected['small_uint'],
            medium_uint=expected['medium_uint'],
            regular_uint=expected['regular_uint'],
            large_uint=large_uint,
            single_precision=expected['single_precision'],
            double_precision=expected['double_precision'],
            flag=expected['flag'],
            device_id=expected['device_id'].encode('utf-8'),
            description=expected['description'].encode('utf-8')
        )

        packet = BasicFrame()
        encoded_data = packet.encode_msg(msg)

        with open('python_basic_types_test_data.bin', 'wb') as f:
            f.write(bytes(encoded_data))

        return True

    except ImportError:
        return True  # Skip if generated code not available

    except Exception as e:
        print_failure_details(
            f"Create test data exception: {type(e).__name__}",
            expected_values={"result": "success"},
            actual_values={"exception": str(e)}
        )
        import traceback
        traceback.print_exc()
        return False


def main():
    """Main test function"""
    print("\n[TEST START] Python Cross-Platform Basic Types Serialization")
    
    success = create_test_data()
    
    status = "PASS" if success else "FAIL"
    print(f"[TEST END] Python Cross-Platform Basic Types Serialization: {status}\n")
    
    return success


if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: tests/cpp/test_cross_platform_basic_types_serialization.cpp
#include "basic_types.sf.hpp"
#include "basic_frame.hpp"
#include <iostream>
#include <fstream>
#include <cstring>

void print_failure_details(const char* label) {
    std::cout << "\n============================================================\n";
    std::cout << "FAILURE DETAILS: " << label << "\n";
    std::cout << "============================================================\n\n";
}

int main() {
    std::cout << "\n[TEST START] C++ Cross-Platform Basic Types Serialization\n";
    
    try {
        BasicTypesBasicTypesMessage msg{};
        // Use values from expected_values.json
        msg.small_int = -42;
        msg.medium_int = -1000;
        msg.regular_int = -100000;
        msg.large_int = -1000000000LL;
        msg.small_uint = 255;
        msg.medium_uint = 65535;
        msg.regular_uint = 4294967295U;
        msg.large_uint = 18446744073709551615ULL;
        msg.single_precision = 3.14159f;
        msg.double_precision = 2.718281828459045;
        msg.flag = true;
        std::strncpy(msg.device_id, "TEST_DEVICE_12345678901234567890", 32);
        msg.description.length = 33;
        std::strncpy(msg.description.data, "Test description for basic types", sizeof(msg.description.data));
        
        uint8_t buffer[512];
        StructFrame::BasicFrameEncodeBuffer encoder(buffer, sizeof(buffer));
        
        if (!encoder.encode(BASIC_TYPES_BASIC_TYPES_MESSAGE_MSG_ID, &msg, 
                            BASIC_TYPES_BASIC_TYPES_MESSAGE_MAX_SIZE)) {
            print_failure_details("Failed to encode message");
            std::cout << "[TEST END] C++ Cross-Platform Basic Types Serialization: FAIL\n\n";
            return 1;
        }
        
        std::ofstream file("cpp_basic_types_test_data.bin", std::ios::binary);
        if (!file) {
            print_failure_details("Failed to create test data file");
            std::cout << "[TEST END] C++ Cross-Platform Basic Types Serialization: FAIL\n\n";
            return 1;
        }
        file.write(reinterpret_cast<const char*>(buffer), encoder.size());
        file.close();
        
        std::cout << "[TEST END] C++ Cross-Platform Basic Types Serialization: PASS\n\n";
        return 0;
        
    } catch (const std::exception& e) {
        print_failure_details(e.what());
        std::cout << "[TEST END] C++ Cross-Platform Basic Types Serialization: FAIL\n\n";
        return 1;
    }
}
